### PR TITLE
bpo-40275: Adding a threading_helper in test.support

### PR DESCRIPTION
**Pull request by shihai1991**
https://bugs.python.org/issue40275

### CODE REVIEW

File: Doc/library/test.rst [hunk 1]
```diff
@@ -1628,3 +1594,48 @@ The module defines the following class:
...
+
+:mod:`test.support.threading_helper` --- Utilities for threading tests 
+======================================================================
+
+.. module:: test.support.threading_helper
+   :synopsis: Support for threading tests.
+
+The :mod:`test.support.threading_helper` module provides support for threading tests.
+
+.. versionadded:: 3.9
```
Comment (vstinner) [inline Doc/library/test.rst]: Hey, master became Python 3.10 :-) Please update your PR ;-)

File: Doc/library/test.rst [hunk 2]
```diff
@@ -1628,3 +1594,48 @@ The module defines the following class:
...
+
+
+.. function:: join_thread(thread, timeout=30.0)
```
Comment (vstinner) [inline Doc/library/test.rst]: 30.0 is outdated: the default is now None, please update the doc.

File: Doc/library/test.rst [hunk 3]
```diff
@@ -1628,3 +1594,48 @@ The module defines the following class:
...
+   Cleanup up threads not specified in *original_values*.  Designed to emit
+   a warning if a test leaves running threads in the background.
+
+
+.. function:: threading_setup()
+
+   Return current thread count and copy of dangling threads.
+
+
+.. function:: wait_threads_exit(timeout=60.0)
```
Comment (vstinner) [inline Doc/library/test.rst]: Same here: the default is now None.

File: Lib/test/support/threading_helper.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
+import contextlib
+import functools
+import _thread
+import threading
+import time
+
+from test import support
+
```
Comment (vstinner) [inline Lib/test/support/threading_helper.py:8]: it's the opportunity to make the code a little bit more PEP 8 compliant: please add an empty line here :-D

File: Lib/test/support/threading_helper.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+# moral equivalent thereof) because a threading.Thread object is still alive
+# until its __bootstrap() method has returned, even after it has been
+# unregistered from the threading module.
+# thread._count(), on the other hand, only gets decremented *after* the
+# __bootstrap() method has returned, which gives us reliable reference counts
+# at the end of a test run.
+
+def threading_setup():
+    return _thread._count(), threading._dangling.copy()
+
```
Comment (vstinner) [inline Lib/test/support/threading_helper.py:24]: PEP 8: add an empty line.

File: Lib/test/test_sys.py [hunk 6]
```diff
@@ -11,6 +9,9 @@
 import sys
 import sysconfig
 import test.support
+from test import support
+from test.support.script_helper import assert_python_ok, assert_python_failure
+from test import threading_helper
```
Comment (vstinner) [inline Lib/test/test_sys.py]: wait, it lives into test.support, not test, no?

File: Doc/library/test.rst (hunk 1 above)
Comment (shihai1991) [reply Doc/library/test.rst]: Oh, got it. Thanks.

File: Lib/test/support/threading_helper.py (hunk 4 above)
Comment (shihai1991) [reply Lib/test/support/threading_helper.py:8]: got it, thanks.

File: Lib/test/test_sys.py (hunk 6 above)
Comment (shihai1991) [reply Lib/test/test_sys.py]: good catch, thanks.

**Review by vstinner — APPROVED** [review]
LGTM, but I would prefer to get @serhiy-storchaka review on this PR since it changes many tests.

**Comment by shihai1991** [comment]
> LGTM, but I would prefer to get @serhiy-storchaka review on this PR since it changes many tests.

Sure, my eyes can't see clealy after changed those tests ; )

**Comment by vstinner** [comment]
Hum, sadly threading is still imported indirectly by subprocess and platform, and so this PR is not enough. But I merge the PR anyway to ease @shihai1991 work, since it's a huge PR and so it's hard to maintain. Overall, it enhances test.support anyway ;-) I will work on following changes to import platform and subprocess lazily.

**Comment by gpshead** [comment]
A question relevant to the test.support APIs this PR introduced: https://discuss.python.org/t/test-harness-for-leakage-detection-does-threading-setup-duplicates-saved-test-environment/17097

**Comment by serhiy-storchaka** [comment]
This PR did not introduce any new API. It is a pure refactoring.

**Pull request merged**